File: examples/NYC/src/GeoSpaceModule.py
from mesa.visualization.ModularVisualization import VisualizationElement
from folium.utilities import image_to_url

from .geo_space import RasterLayer, VectorLayer


class GeoSpaceModule(VisualizationElement):
    package_includes = ["leaflet.js"]
    local_includes = ["src/GeoSpaceModule.js"]

    def __init__(self, portrayal_method, view=[0, 0], zoom=10, map_height=500, map_width=500):
        super().__init__()

        self.portrayal_method = portrayal_method
        self.map_height = map_height
        self.map_width = map_width
        self.view = view
        new_element = "new GeoSpaceModule({}, {}, {}, {})"
        new_element = new_element.format(view, zoom, map_width, map_height)
        self.js_code = "elements.push(" + new_element + ");"

    def render_layers(self, model):
        layers = {
            "rasters": [image_to_url(layer.values.transpose([1, 2, 0]))
                        for layer in model.grid.layers if isinstance(layer, RasterLayer)],
            "vectors": [layer.__geo_interface__ for layer in model.grid.layers if isinstance(layer, VectorLayer)],
            # longlat to latlong
            "bounds": [list(reversed(model.grid.bounds[0])), list(reversed(model.grid.bounds[1]))]
        }
        return layers

    def render(self, model):
        featurecollection = dict(type="FeatureCollection", features=[])
        for _, agent in enumerate(model.grid.agents):
            shape = agent.__geo_interface__()
            portrayal = self.portrayal_method(agent)
            for key, value in portrayal.items():
                shape["properties"][key] = value
                featurecollection["features"].append(shape)
        return featurecollection

File: examples/NYC/src/GeoSpaceModule.js
var GeoSpaceModule = function (view, zoom, map_width, map_height) {
    // Create the map tag:
    var map_tag = "<div style='width:" + map_width + "px; height:" + map_height + "px;border:1px dotted' id='mapid'></div>"
    // Append it to body:
    var div = $(map_tag)[0]
    $('#elements').append(div)

    // Create Leaflet map and agent layer
    var Lmap = L.map('mapid', {zoomSnap: 0.1}).setView(view, zoom)
    var agentLayer = L.geoJSON().addTo(Lmap)

    this.renderLayers = function (layers) {
        layers.rasters.forEach(function (layer) {
            L.imageOverlay(layer, layers.bounds).addTo(Lmap)
        })
        layers.vectors.forEach(function (layer) {
            L.geoJSON(layer).addTo(Lmap)
        })
        Lmap.fitBounds(layers.bounds)
    }

    this.render = function (data) {
        agentLayer.remove()
        agentLayer = L.geoJSON(data, {
            onEachFeature: PopUpProperties,
            style: function (feature) {
                return {color: feature.properties.color};
            },
            pointToLayer: function (feature, latlang) {
                return L.circleMarker(latlang, {radius: feature.properties.radius, color: feature.properties.color});
            }
        }).addTo(Lmap)
    }

    this.reset = function () {
        agentLayer.remove()
    }
}

function PopUpProperties(feature, layer) {
    var popupContent = '<table>'
    if (feature.properties) {
        for (var p in feature.properties) {
            popupContent += '<tr><td>' + p + '</td><td>' + feature.properties[p] + '</td></tr>'
        }
    }
    popupContent += '</table>'
    layer.bindPopup(popupContent)
}